### PR TITLE
add rust runnable

### LINKED ISSUE #16160: Use lsp runnables
### Description

as of right now rust tests are being detected by custom code. It would probably be easier to use rust-analyzer for that. This would give the ide context like in which mod it is. I had issues where no test was run or multiple at the same time. 

Rust has implemented this as `experimental/runnables`:
https://github.com/rust-lang/rust-analyzer/blob/master/crates/ide/src/runnables.rs

But the upstream LSP issue was closed for lack of interest:
- https://github.com/microsoft/language-server-protocol/issues/944

### PR DESCRIPTION
**Pull request by frederik-uni**
# THIS IS A DRAFT!!

## Issues
- how to sent/convert location?
- how to update the detected tests(the runnable.scm detection detects tests that aren't tests & the commands are faulty. Use find runnables for each file a test was detected in correct issues)?

## Related
Closes #16160

Release Notes:

- Fixed Rust test detection

### CODE REVIEW

**Review by SomeoneToIgnore — COMMENTED** [review]
Huh, interesting how @P1n3appl3 , me and now you think of the same actively.

On my side, I've made a stab at https://github.com/zed-industries/zed/compare/main...kb/r-a-tasks trying to find more big issues along the way to make it work.

> how to sent/convert location?

Is it the `lsp::LocationLink` to `project::LocationLink` conversion meant?
`location_links_from_proto`'s signature was changes so you must have also noticed a somewhat specific `location_links_from_lsp` — I think a part of it can be extracted to share with this conversion need.

On the other hand, this (ergo, all `proto.rs` adjustments) might not be needed in the end: rust-analyzer's runnables are Zed's `TaskTemplate` (fully resolved, as a bonus).

So all we need to do instead is an lsp -> TaskTemplate conversion and whatever existing to_proto conversions of templates. 


> how to update the detected tests

It does not matter that much in the context of `TaskTemplate`s, but to be precise: rust-analyzer's tasks are not only tests, but generic `cargo` tasks and even `shell` fwiw?

https://github.com/rust-lang/rust-analyzer/blob/fb133c8c8064d9f18f19eae8721347872f1679e4/crates/rust-analyzer/src/lsp/ext.rs#L458-L463

One way or another, those task templates has to be returned when 


* an editor queries for its task data updated:

 https://github.com/zed-industries/zed/blob/14772f4e8641edd628b6654e707cf5e5532fd6a2/crates/editor/src/editor.rs#L12295

This gets green arrows to editor's gutter and is usually updated on file updates (`multi_buffer::Event::Reparsed`) or new editor regions (new multi buffer excerpt added).

Note that this excludes any position-related changes (e.g. selections), assuming tasks are queried for the file (each entire file, for multi buffer).

Looking at https://github.com/rust-lang/rust-analyzer/blob/fb133c8c8064d9f18f19eae8721347872f1679e4/crates/rust-analyzer/src/handlers/request.rs#L954-L959

it seems that `offset` is used for filtering out runnables only, so we can never send that and get runnables for the entire file always.

We might want to do this with a debounce 
https://github.com/zed-industries/zed/blob/14772f4e8641edd628b6654e707cf5e5532fd6a2/crates/editor/src/editor.rs#L5083-L5087


* a task modal (`task: spawn` opens it)

There, my attempt had faced quite a nasty case, where all tree-sitter based tasks were derived synchronously which is definitely not the model for our LSP + remote interactions.

https://github.com/zed-industries/zed/compare/main...kb/r-a-tasks#diff-7eeacbb475dccf579442ccc356338bf6bf62f4515088508eca227374579546baR226

notices the issue.

Overall, we seem to be ok as most of our callers up the stack are UI or action runners and have an ability to do `cx.spawn(...).detach()` up the stack.
It needs some plumbing though.

-----------------------------

I think we should close this PR to keep the queue cleaner, as its proto part is not relevant in the end design (explained above), and its LSP-related part is somewhat strange as could've just taken r-a's model exactly instead of paraphrasing.

https://github.com/zed-industries/zed/compare/main...kb/r-a-tasks is somewhat closer to impl, but is still eons away, hence not opening a PR yet.

Main points to mention:

* async LSP queries in currently synchronous task-related API (that uses tree-sitter).

* more work on ergonomics: ability to turn off both LSP and tree-sitter tasks separately, in the settings (the tree-sitter part starts https://github.com/zed-industries/zed/blob/14772f4e8641edd628b6654e707cf5e5532fd6a2/crates/language/src/language_settings.rs#L1006 [1] and could be a nice standalone PR!)

* `refresh_runnables` + lsp -> `TaskTemplate` conversion

Seems to be quite minor compared to the rest

* Working in remote projects (ssh and collab)

One thing is a relatively "marginal" "expand macros" command, another thing is a large nice feature not working — it better be.

I'm currently working on the last issue as the hardest, in https://github.com/zed-industries/zed/compare/main...kb/remote-server-id [2]
it also has some async-related issues, but feels halfway there.

My next few weeks' schedule would be quite tough, so I'm not sure how fast this all goes, but this is what I'm actively working on when I can.

Let me know if I can make it more collaborative-friendly if you want to participate or help otherwise.
I think there are at least [1] and [2] roads fully standalone and very important to work on, if the current topic seems like too much at this point.
